### DISCUSSION TITLE
Lesson 7: Error While running a Test

### DISCUSSION THREAD

**Discussion by Ifra001** [Cyfrin/foundry-full-course-cu#1830] [General]
I am running a test and it is giving me this error
```
[FAIL. Reason: panic: arithmetic underflow or overflow (0x11)] testFundUpdatesDataStructure() (gas: 31857)
```
This is my Code for the test, I think it is exactly the same as Patrick had written it (Instead of fundMe, my contract is called "donations")
```solidity
  function testFundUpdatesDataStructure() public {
        vm.prank(USER); //The next TX will be sent by user
        donations.fund{value: SEND_VALUE}(); //SEND_VALUE is 0.1 ether

        uint256 amountFunded = donations.getAddressToAmountFunded(USER);
        assertEq(amountFunded,SEND_VALUE);
    }
```
Please Help me :(

**Comment by EngrPips** [comment]
Hello @DontMind-me, Please run your test with the `-vvvv` flag and show the test output.

**Reply by Ifra001** [reply]
Here you go
```solidity
Ran 1 test for test/DonationsTest.t.sol:DonationTest
[FAIL. Reason: panic: arithmetic underflow or overflow (0x11)] testFundUpdatesDataStructure() (gas: 31857)
Traces:
  [2904082] DonationTest::setUp()
    ├─ [1448328] → new DeployDonations@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 7013 bytes of code
    ├─ [1396997] DeployDonations::run()
    │   ├─ [926036] → new <unknown>@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [377261] → new <unknown>@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← [Return] 1107 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] 2230 bytes of code
    │   ├─ [359] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   ├─ [0] VM::startBroadcast()
    │   │   └─ ← [Return] 
    │   ├─ [404008] → new Donations@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   └─ ← [Return] 1906 bytes of code
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] Donations: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496]
    ├─ [0] VM::deal(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D], 10000000000000000000 [1e19])
    │   └─ ← [Return] 
    └─ ← [Stop] 

  [31857] DonationTest::testFundUpdatesDataStructure()
    ├─ [0] VM::prank(user: [0x6CA6d1e2D5347Bfab1d91e883F1915560e09129D])
    │   └─ ← [Return] 
    ├─ [14992] Donations::fund{value: 100000000000000000}()
    │   ├─ [8993] 0x90193C961A926261B756D1E5bb255e67ff9498A1::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 200000000000 [2e11], 1, 1, 1
    │   └─ ← [Revert] panic: arithmetic underflow or overflow (0x11)
    └─ ← [Revert] panic: arithmetic underflow or overflow (0x11)

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 148.92ms (4.25ms CPU time)

Ran 1 test suite in 3.39s (148.92ms CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/DonationsTest.t.sol:DonationTest
[FAIL. Reason: panic: arithmetic underflow or overflow (0x11)] testFundUpdatesDataStructure() (gas: 31857)

Encountered a total of 1 failing tests, 0 tests succeeded
```

**Reply by EngrPips** [reply]
Thank you. Please show the `fund` function inside your `Donations` contract.

**Reply by Ifra001** [reply]
@EngrPips Here is my ```fund``` function
```solidity
    function fund() public payable {
        // require(msg.value.getConversionRate() >= MINIMUM_USD, "Didn't Send Enought ETH!!");
        if (msg.value.getConversionRate(s_pricefeed) < MINIMUM_USD) { //Minimum USD is 5 dollars
            revert Donations__NotEnoughEth();
        }
        s_funders.push(msg.sender);
        s_addressToAmountFunded[msg.sender] =
            s_addressToAmountFunded[msg.sender] +
            msg.value;
    }
```

**Reply by EngrPips** [reply]
Thanks, Please show the `getConversionRate` inside your `PriceConverter` library.

**Reply by Ifra001** [reply]
I sent the whole ```PriceConverter```  library 
```solidity
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

library PriceConverter {
    function getPrice(AggregatorV3Interface pricefeed) public view returns (uint256) {
        (, int256 answer, , , ) = pricefeed.latestRoundData();
        return uint256(answer ** 10000000000);
    }

    function getConversionRate(
        uint256 ethAmount,
        AggregatorV3Interface pricefeed
    ) public view returns (uint256) {
        uint256 ethPrice = getPrice(pricefeed);
        uint256 ethInUSD = (ethPrice * ethAmount) / 1000000000000000000;
        return ethInUSD;
    }
}
```

**Reply by EngrPips** [reply]
Thanks. My `getConversionRate` function is slightly different from yours. See mine below

```js
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.23;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

library PriceConverter {
    // ETHUSD PRICEFEED ON SEPOLIA => 0x694AA1769357215DE4FAC081bf1f309aDC325306

    function getConversionRate(uint256 _ethValueToConvert, address _priceFeedAddress) internal view returns(uint256) {
        uint256 price = getPrice(_priceFeedAddress);
        uint256 receivedValueInUSD = _ethValueToConvert * (price * 10 ** 10);
        return receivedValueInUSD;
    }

    function getPrice(address _priceFeedAddress) internal view returns(uint256) {
        AggregatorV3Interface priceFeedContract = AggregatorV3Interface(_priceFeedAddress);
        (,int answer,,,) = priceFeedContract.latestRoundData();
        uint256 price = uint256(answer);
        return price;
    }    
}
```

**Reply by EngrPips** [reply]
> I sent the whole `PriceConverter` library
> 
> ```solidity
> pragma solidity ^0.8.18;
> 
> import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
> 
> library PriceConverter {
>     function getPrice(AggregatorV3Interface pricefeed) public view returns (uint256) {
>         (, int256 answer, , , ) = pricefeed.latestRoundData();
>         return uint256(answer ** 10000000000);
>     }
> 
>     function getConversionRate(
>         uint256 ethAmount,
>         AggregatorV3Interface pricefeed
>     ) public view returns (uint256) {
>         uint256 ethPrice = getPrice(pricefeed);
>         uint256 ethInUSD = (ethPrice * ethAmount) / 1000000000000000000;
>         return ethInUSD;
>     }
> }
> ```

Can you please put a `console.log()` after this line `uint256 ethInUSD = (ethPrice * ethAmount) / 1000000000000000000;` in your
`getConversionRate` function so that we can see what  `ethInUSD` resulted to.

**Reply by Ifra001** [reply]
> Thanks. My `getConversionRate` function is slightly different from yours. See mine below
> 
> ```js
> // SPDX-License-Identifier: UNLICENSED
> pragma solidity 0.8.23;
> 
> import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
> 
> library PriceConverter {
>     // ETHUSD PRICEFEED ON SEPOLIA => 0x694AA1769357215DE4FAC081bf1f309aDC325306
> 
>     function getConversionRate(uint256 _ethValueToConvert, address _priceFeedAddress) internal view returns(uint256) {
>         uint256 price = getPrice(_priceFeedAddress);
>         uint256 receivedValueInUSD = _ethValueToConvert * (price * 10 ** 10);
>         return receivedValueInUSD;
>     }
> 
>     function getPrice(address _priceFeedAddress) internal view returns(uint256) {
>         AggregatorV3Interface priceFeedContract = AggregatorV3Interface(_priceFeedAddress);
>         (,int answer,,,) = priceFeedContract.latestRoundData();
>         uint256 price = uint256(answer);
>         return price;
>     }    
> }

In your ```getConversionRate``` function, you did not divide the ```ethValuetoConvert``` and ```price``` with 10 raised to the power of 10, rather you multiplied it, it that where I did something wrong?

**Reply by EngrPips** [reply]
Try to change that to be like mine and re-run the test, let's see if it pass


> > I sent the whole `PriceConverter` library
> > ```solidity
> > pragma solidity ^0.8.18;
> > 
> > import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
> > 
> > library PriceConverter {
> >     function getPrice(AggregatorV3Interface pricefeed) public view returns (uint256) {
> >         (, int256 answer, , , ) = pricefeed.latestRoundData();
> >         return uint256(answer ** 10000000000);
> >     }
> > 
> >     function getConversionRate(
> >         uint256 ethAmount,
> >         AggregatorV3Interface pricefeed
> >     ) public view returns (uint256) {
> >         uint256 ethPrice = getPrice(pricefeed);
> >         uint256 ethInUSD = (ethPrice * ethAmount) / 1000000000000000000;
> >         return ethInUSD;
> >     }
> > }
> > ```
> 
> Can you please put a `console.log()` after this line `uint256 ethInUSD = (ethPrice * ethAmount) / 1000000000000000000;` in your `getConversionRate` function so that we can see what `ethInUSD` resulted to.

You can also still do this and let's see what is the value that your implementation is resulting to which is causing the `overflow`/`underflow` issue.

**Reply by Ifra001** [reply]
I solved it! It was because of the math. I think I divided and multiplied incorrectly resulting in my tests failing. Thank you so much for your help @EngrPips . Math in Solidity really is confusing and its hard to check for me

**Reply by EngrPips** [reply]
Yeah, Math in solidity can be tricky at first but with time it will make more sense and becomes more clear. Keep crushing it.
